tools/autograd/templates/VariableType.cpp

```diff
@@ -104,7 +104,14 @@ Variable & VariableType::checked_cast(const Type & type, const Tensor & t, const
     runtime_error("Expected a Tensor of type %s but found an undefined Tensor for argument #%d '%s'",
         type.toString(), pos, name);
   }
-  if (&t.type() != &type && &t.type() != &type.toBackend(toSparse(t.type().backend()))) {
+
+  // Not all types have a matching sparse type
+  Type* sparse_type = &type.toScalarType(ScalarType::Undefined);
+  try {
+    sparse_type = &type.toBackend(toSparse(t.type().backend()));
+  } catch (std::runtime_error &e) { }
```

Comment:
`try {} catch {}` isn't free when exceptions are thrown, so it would be better to use `if`s for that purpose.

Reply:
I didn't find how to check whether a type has sparse support. The only way I saw possible is to check whether `context.type_registry[backend][type]` is nullptr. Does that sound good enough?

Reply:
That seems to be a very hacky way. We should try to come up with something better (perhaps we should have `toBackendOpt` that returns a pointer instead of a reference, and this can be `null`)

Reply:
Yeah I agree. Something better would be:
1. a `at::Context::getTypeOpt` which doesn't throw
2. a `at::Type::toBackend` and `at::Type::toScalarType` which don't throw.

In terms of return value, would you prefer a reference to undefined, or a null-able pointer?

Reply:
Since we're treating undefined as a first class type I think it would make sense to indicate absence as a `nullptr` (especially that it can only happen in the `Opt` overload where you're explicitly asking for it). However, I'm not sure how well does this fit for the current/future vision for ATen, so I'll leave the final decision up to @gchanan or @colesbury

Reply:
I think returning a nullptr is more idiomatic C++, and we like idiomatic C++. Undefined type is purely there to solve the "we don't want segfaults when forget to check for undefined", but we (I?) hate it, see #4274 and #4215.

Reply:
Sorry, can we take a step back?  What command are you running and what is the crash?

Reply:
@gchanan In fact I don't know how to trigger this! I met this problem because of some other changes I made to pytorch about its half tensor support. But this is still a bug I think.

Reply:
it's difficult to make an informed decision without reproduction; the whole thing seems strange because I would just expect you'd get an exception at the end of whatever you are doing, which seems like the correct behavior.

Reply:
Yes I will still get an exception, but with the wrong error message.

Reply:
Expected to get "Expected object of type %s but found type %s for argument #%d '%s'". But instead I will get something like SparseCUDAHalfTensor not enabled.

Reply:
I think the central issue is without the context it's hard to know why you expected to get "Expected object of type %s but found type %s for argument #%d '%s'".

Reply:
I don't think we want to do it this way. This will add overhead to every PyTorch function.

We should fix the unpacking logic (see #4547). I'm not sure it's worth putting too much effort into fixing error messages in a function thats likely to change soon.

Reply:
Wasn't aware of #4547 but that one will solve my problem.

Reply:
I think I can close this now and wait for the coming change.